pkg/gpu/cuda/env.go

```diff
@@ -0,0 +1,117 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package cuda
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+)
+
+const cudaVisibleDevicesEnvVar = "CUDA_VISIBLE_DEVICES"
+
+// GetVisibleDevicesForProcess modifies the list of GPU devices according to the
+// value of the CUDA_VISIBLE_DEVICES environment variable for the specified
+// process. Reference:
+// https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
+// Invalid device indexes are ignored, and anything that comes after that is
+// invisible, following the spec: "If one of the indices is invalid, only the
+// devices whose index precedes the invalid index are visible to CUDA
+// applications." If an invalid index is found, an error is returned together
+// with the list of valid devices found up until that point.
+func GetVisibleDevicesForProcess(systemDevices []nvml.Device, pid int, procfs string) ([]nvml.Device, error) {
+	cudaVisibleDevices, err := kernel.GetProcessEnvVariable(pid, cudaVisibleDevicesEnvVar, procfs)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get env var %s for process %d: %w", cudaVisibleDevicesEnvVar, pid, err)
+	}
+
+	return getVisibleDevices(systemDevices, cudaVisibleDevices)
+}
+
+// getVisibleDevices processes the list of GPU devices according to the value of
+// the CUDA_VISIBLE_DEVICES environment variable
+func getVisibleDevices(systemDevices []nvml.Device, cudaVisibleDevices string) ([]nvml.Device, error) {
+	if cudaVisibleDevices == "" {
+		return systemDevices, nil
+	}
+
+	var filteredDevices []nvml.Device
+	visibleDevicesList := strings.Split(cudaVisibleDevices, ",")
+
+	for _, visibleDevice := range visibleDevicesList {
+		var matchingDevice nvml.Device
+		var err error
+		if strings.HasPrefix(visibleDevice, "GPU-") {
+			matchingDevice, err = getDeviceWithMatchingUUIDPrefix(systemDevices, visibleDevice)
+			if err != nil {
+				return filteredDevices, err
+			}
+		} else if strings.HasPrefix(visibleDevice, "MIG-GPU") {
+			// MIG (Multi Instance GPUs) devices require extra parsing and data
+			// about the MIG instance assignment, which is not supported yet.
+			return filteredDevices, fmt.Errorf("MIG devices are not supported")
```

Comment:
It looks weird to me too, but the problem is how to replicate the behavior documented by NVIDIA here. Any error in the list doesn't cause an actual error, but just ignores that device and everything that follows. I think in this case it's better to return error + partial value, that way we can log the error in case it happens (and we also can check in case it's a bug in our code) but still replicate the same logic they have in the runtime.

Reply:
why do you need that error at all?

Reply:
At least while we start testing this, I'd like to have logs and possibly telemetry on failures to parse this environment variable, so if it happens too much or if we see weird GPU assignments we can investigate.

Reply:
ok i think we can replace those errors with telemetry metrics in the following PRs it will make the function a bit more clear.